File: hw6/161044121/GTUSet.cpp
//the functions are almost like vector functions but in the insert the set class doesn't allow duplicate element
//beside these the changes are mainly in the functions parameter and in very less amount

#include "GTUSet.h"
namespace swr{
    template<class T>
    void GtuSet<T>::expand(){
        try{
            int newSize=this->capacity*2;
            if(newSize>this->max){throw newSize;}
            shared_ptr<T> temp(new T[newSize]{0},[](T* ptr){delete[] ptr;});
            for(int i=0; i<this->used;i++){
                temp.get()[i]=this->array.get()[i];
            }
        }
        catch(int newSize){
                cout<<"max capacity reached!!!\n";
        }
    }
    template<class T>
    bool GtuSet<T>::empty()const{
        if(this->used==0){
            return true;
        }
        return false;
    }
    template<class T>
    int GtuSet<T>::size()const{
        return this->used;
    }
    template<class T>
    int GtuSet<T>::max_size()const{
        return this->max;
    }
    template<class T>
    void GtuSet<T>::insert(const T& x){
        int flag=true;
        if(this->used == this->capacity){expand();}
        else{
            for(int i=0;i<this->used;i++){
                if(this->array.get()[i]==x){
                    flag=false;
                    break;
                }
            }
            if(flag==true){
                this->array.get()[this->used]=x;
                this->used++;
            }
        }
    }
    template<class T>
    void GtuSet<T>::erase(const T& value){ 
       int flag=false,i;
       for( i=0; i<this->used; i++){
           if(this->array.get()[i]==value){
               flag=true;
               break;
           }  
        }
        cout<<i<<endl;
        if(flag==true){
            for(int j=i; j<this->used;j++){
                this->array.get()[j]=this->array.get()[j+1];
            }
        }
        
        
    }
    template<class T>
    void GtuSet<T>::erase(T* pos){
        for(auto i = pos; i<this->end(); i++){
            *i=*(i+1);
        }
        this->used--;
    }
    template<class T>
    void GtuSet<T>::clear(){
        this->capacity=10;
        this->used=0;
        shared_ptr<T> temp(new T[this->capacity],[](T* ptr){delete[] ptr;});
        this->array=temp;
    }
    template<class T>
    T* GtuSet<T>::begin(){
        return &(*(this->array));
    }
    template<class T>
    T* GtuSet<T>::end(){
        return &(this->array.get()[this->used]);
    }
    template<class T>
    T& GtuSet<T>::operator[](int index){
        if(index>=this->used){
            cout<<"invalid index"<<endl;
        }
        return this->array.get()[index];
    }
}

File: hw8/161044121/GTUSet.java
import Gtu.util.GTUIterator;

public class GTUSet<T> extends AbstractContainer<T> {
    public boolean empty(){
        if(used==0) return true;
        else return false;
    }
    
    public int size(){
        return used;
    }
    public int max_size(){
        return max;
    }
    public boolean insert(T element){
        if(used==capacity){
            capacity+=10;
            if(capacity>max){
                if(max-capacity==0)
                    return false;
                else
                    capacity=max-capacity;
            }
            T[] temp=(T[]) new Object[capacity];
            for(int i=0; i<used; i++){
                temp[i]=array[i];
            }
            array=(T[]) new Object[capacity];
            for(int i=0; i<used; i++){
                array[i]=temp[i];
            }
        }
        try {
            if(isPresent(element)==false){
                array[used++]=element;
                return true;
            }
        } catch (Exception e) {
            System.out.println("error while inserting element");
        }
        return false;
    }
    public boolean erase(T element){
        try {
            if(isPresent(element)){
                int i;
                for( i=0; i<used; i++){
                    if(array[i]==element)
                        break;
                }
                for(int j=i; j<used; j++){
                    array[j]=array[j+1];
                }
                used-=1;
                return true;
            }
            throw new Exception("invalid element to erase");
        } catch (Exception e) {
            System.out.println(e);
            return false;
        }
    }
    public void clear(){
        used=0;
        capacity=10;
        array=(T[])new Object[capacity];
    }
    public boolean contains(Object o){
        for(int i=0; i<used; i++){
            if((Object)array[i]==o){
                return true;
            }
        }
        return false;
    }
    public class setIter<T> implements GTUIterator<T>{

        private int index;
        public setIter(){
            index=0;
        }
        public boolean hasNext() {
            if(index < used){
                return true;
            }
            return false;
        }
        public T next() {
            return (T) array[index++];
        }

    }
    public GTUIterator<T> iterator(){
        return new setIter<T>();
    }
    private boolean isPresent(T key){
        for(int i=0; i<used; i++){
            if(array[i]==key){
                return true;
            }
        }
        return false;
    }
}